### ISSUE TITLE
How to improve DETR on small objects (mask detection) ?

### ISSUE THREAD

**Issue by jeromen7** [facebookresearch/detr#216]
Hello,

I am currently working on a mask detector using DETR and it is working pretty well :

![image](https://user-images.githubusercontent.com/45367085/91441360-5e0b4700-e870-11ea-9921-f5022353266c.png)

But the results are very poor when it comes to detecting masks in a large crowd. Here is an example : 

![image](https://user-images.githubusercontent.com/45367085/91440212-88f49b80-e86e-11ea-8984-90b1a03d4bf0.png)
As you can see, the detection boxes are very random, and the results are pretty bad.
How can I improve that ?

- I was thinking about dividing the image in 4 parts, zoom in on each part and detect the masks, and the reassemble the image but it seems complicated and it will increase the inference time a lot.

- The other solution I imagined was getting a lot of crowd images like the one above, labelise them and add them to my dataset. But if DETR is by definition not working well on small objects like it is said on the paper, maybe this is not a good solution...

Would you have any suggestions that my help ?
Thank you all !

**Comment by lessw2020** [comment]
Hi @jeromen7 
Great work on the mask detector!
To your question, dividing the image into quadrants and doing inference on each would play to detrs strength with larger objects and would be the best approach to try imo.
I do a lot of subset, rotate, rescale stuff for my work with detr and once you get the manipulations down it works really well.
You might also try with ciou loss, but that would be more speculative how much it would help with small objects. The implementation is in my repo though so easy to try.
Hope that helps

**Comment by Epiphqny** [comment]
@jeromen7 Have you increase the number of queries of the model? It seems there are more than 100 objects in the test image.

**Comment by woctezuma** [comment]
I have roughly counted the number of green boxes, and it is close to 100, if not exactly 100. I counted at least 80.

**Comment by jeromen7** [comment]
Thank you all for your help !
I will try the solution to split the image in quadrants, and I will come back to tell you if it worked.

**Comment by jeromen7** [comment]
Here is the result : 
![image](https://user-images.githubusercontent.com/45367085/92131841-b9a47a00-ee06-11ea-87a6-3d745a53dcd0.png)
As you can see it works much better, still not perfect but it definitely improved. However it also takes more inference time, as expected ...
I am closing the issue, thank you again @lessw2020 @Epiphqny @woctezuma  !

**Comment by meghalD** [comment]
> I have roughly counted the number of green boxes, and it is very close to 100. I counted at least 80, likely more than 90.
> 
> > I was thinking about dividing the image in 4 parts, zoom in on each part and detect the masks, and the reassemble the image
> 
> Do it. Or train the model from scratch with more than 100 query slots
> 
> The latter sounds more complicated, especially if you have to use 1000 query slots, or who knows, 10,000 query slots.
> 
> References:
> 
> > Note that if your dataset has a substantially different average number of objects per image than coco, you might need to adjust the number of object queries (--num_queries) It should be strictly higher than the max number of objects you may have to detect, and it's good to have some slack (in coco we use 100, the max number of objects in a coco image is ~70)
> 
> [#9 (comment)](https://github.com/facebookresearch/detr/issues/9#issuecomment-635357693)
> 
> > I'd recommend trying to fine-tune the class head, while starting from a pre-trained encoder/decoder. You'll have to keep the 100 queries if you do that, but unless you're after very marginal speed improvement it shouldn't hurt.
> 
> [#9 (comment)](https://github.com/facebookresearch/detr/issues/9#issuecomment-639456002)
> 
> > Unless you literally have 700 items in each image, do not change num_queries and keep it at 100, this will give you 100 proposed boxes for each item. Changing num_queries will result in retraining the whole transformer, which is costly.
> 
> [#9 (comment)](https://github.com/facebookresearch/detr/issues/9#issuecomment-640480904)
> 
> > If you're fine-tuning, I don't recommend changing the number of queries on the fly, it is extremely unlikely to work out of the box. In this case you're probably better off retraining from scratch (you can change the `--num_queries` arg from our training script).
> 
> [#9 (comment)](https://github.com/facebookresearch/detr/issues/9#issuecomment-636407752)
> 
> > 500 queries is quite high, it's fairly possible that the default hyper-parameters are not optimal for this setting.
> 
> [#9 (comment)](https://github.com/facebookresearch/detr/issues/9#issuecomment-645521166)

Can we finetune our model on less no. of queries such as 30/40 for new data ?  I have already converted my data to coco format , ran for maxDet 100. Now I want to check for less queries. How should I modify the code ?

**Comment by woctezuma** [comment]
> Now I want to check for less queries. How should I modify the code ?

I would keep the the maximal number of queries.

Instead, I would threshold the confidence score to find the N best detections, where N<100.

**Comment by FinnMers** [comment]
Look for "Sliced aided hyper inference - SAHI" on Github! This might help! They added detectron2 support, maybe you can implement DETR also somehow.

**Comment by soltanianzadeh** [comment]
Hello. I'm trying to use DETR for small object detection. My images show a crowded region of ~300 small objects (boxes of size ~10x10 pixels) that I'm trying to identify. I played around with the training parameters and have managed to get it to work on my dataset. The problem I have is that I get a lot of significantly overlapping boxes, and thresholding based on score values does not help. I'd appreciate any tips on how I can improve the results.

**Comment by djramakrishna** [comment]
Hi @soltanianzadeh I'm working on something similar, where I've modified DETR to detect small objects (just 1 class btw) (~15x15pixels) in a image resolution of around 1200x1200 pixels. The bounding box loss goes to near zero but the predictions during the training does not match the ground truth in most of the epochs. I did modify the GIOU with CIOU which stabilized the training, but still the predictions does not match up exactly. So, I just wanted to know how you got it to work with smaller objects. Looking forward to your reply. Thank you!
